File: backend/services/insumos/Medicamentos.py
// src/services/insumosMedicamentosService.ts
export type Medicamento = {
  id: string;
  nombre: string;
  principio: string;
  lote: string;
  vencimiento: string;
  laboratorio: string;
  tipo: "Vacuna" | "Antibiótico" | "Antiparasitario" | "Desinfectante" | "Otro";
  condiciones: string;
  proveedor: string;
  stock: number;
  unidad: string;
  observaciones: string;
};

const STORAGE_KEY = "insumos_medicamentos";

function getAll(): Medicamento[] {
  const raw = localStorage.getItem(STORAGE_KEY);
  return raw ? JSON.parse(raw) : [];
}

function saveAll(lista: Medicamento[]) {
  localStorage.setItem(STORAGE_KEY, JSON.stringify(lista));
}

export function getMedicamentos() {
  return getAll();
}

export function addMedicamento(med: Omit<Medicamento, "id">) {
  const nuevo: Medicamento = { ...med, id: crypto.randomUUID() };
  const medicamentos = getAll();
  medicamentos.push(nuevo);
  saveAll(medicamentos);
  return nuevo;
}

export function updateMedicamento(id: string, med: Omit<Medicamento, "id">) {
  const medicamentos = getAll();
  const idx = medicamentos.findIndex(m => m.id === id);
  if (idx >= 0) {
    medicamentos[idx] = { ...med, id };
    saveAll(medicamentos);
    return medicamentos[idx];
  }
  throw new Error("Registro no encontrado");
}

export function deleteMedicamento(id: string) {
  const medicamentos = getAll().filter(m => m.id !== id);
  saveAll(medicamentos);
}

File: frontend/src/services/insumos/Medicamentos.ts
// src/services/insumos/Medicamentos.ts
import API_BASE from "../../config/api";

export type Medicamento = {
  id: number;
  nombre: string;
  principio: string;
  lote: string;
  vencimiento: string;
  laboratorio: string;
  tipo: "Vacuna" | "Antibiótico" | "Antiparasitario" | "Desinfectante" | "Otro";
  condiciones: string;
  proveedor: string;
  stock: number;
  unidad: string;
  observaciones: string;
};

export type NuevoMedicamento = Omit<Medicamento, "id">;

const EMPRESA_ID = 1;
const GRANJA_ID = 1;

function mapApiToMedicamento(api: any): Medicamento {
  return {
    id: api.id,
    nombre: api.nombre,
    principio: api.principio ?? "",
    lote: api.lote ?? "",
    vencimiento: api.vencimiento ?? "",
    laboratorio: api.laboratorio ?? "",
    tipo: api.tipo,
    condiciones: api.condiciones ?? "",
    proveedor: api.proveedor ?? "",
    stock: api.stock,
    unidad: api.unidad,
    observaciones: api.observaciones ?? "",
  };
}

export async function getMedicamentos(): Promise<Medicamento[]> {
  const res = await fetch(
    `${API_BASE}/insumos/medicamentos/?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`
  );
  if (!res.ok) {
    throw new Error("Error al obtener medicamentos");
  }
  const data = await res.json();
  return data.map(mapApiToMedicamento);
}

export async function addMedicamento(
  med: NuevoMedicamento
): Promise<Medicamento> {
  const payload = {
    empresa_id: EMPRESA_ID,
    granja_id: GRANJA_ID,
    nombre: med.nombre,
    principio: med.principio,
    lote: med.lote,
    vencimiento: med.vencimiento || null,
    laboratorio: med.laboratorio,
    tipo: med.tipo,
    condiciones: med.condiciones,
    proveedor: med.proveedor,
    stock: med.stock,
    unidad: med.unidad,
    observaciones: med.observaciones,
  };

  const res = await fetch(`${API_BASE}/insumos/medicamentos/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error("Error al agregar medicamento");
  }
  const data = await res.json();
  return mapApiToMedicamento(data);
}

export async function updateMedicamento(
  id: number,
  med: NuevoMedicamento
): Promise<Medicamento> {
  const payload = {
    nombre: med.nombre,
    principio: med.principio,
    lote: med.lote,
    vencimiento: med.vencimiento || null,
    laboratorio: med.laboratorio,
    tipo: med.tipo,
    condiciones: med.condiciones,
    proveedor: med.proveedor,
    stock: med.stock,
    unidad: med.unidad,
    observaciones: med.observaciones,
  };

  const res = await fetch(
    `${API_BASE}/insumos/medicamentos/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }
  );
  if (!res.ok) {
    throw new Error("Error al actualizar medicamento");
  }
  const data = await res.json();
  return mapApiToMedicamento(data);
}

export async function deleteMedicamento(id: number): Promise<void> {
  const res = await fetch(
    `${API_BASE}/insumos/medicamentos/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    { method: "DELETE" }
  );
  if (!res.ok) {
    throw new Error("Error al eliminar medicamento");
  }
}
